modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy

```
@@ -0,0 +1,123 @@
+ */
+@ToString
+@EqualsAndHashCode
+@CompileStatic
+class Architecture {
+
+/*
+ * example of notation in process: arch 'linux/x86_64', target: 'haswell'
+ * example of notation in config:  arch = [name: 'linux/x86_64', target: 'haswell']
+ * 
+ * where dockerArch = 'linux/x86_64'
+ *       spackArch = target ?: arch  // plus some validation for Spack syntax
+ * 
+ *       platform = 'linux'
+ *       arch = 'x86_64'
+ *       target = 'haswell'
+ * 
+ * [alternate example: 'arch linux/arm/v8', where platform = 'linux' and arch = 'arm/v8']
+ */
+    // used in Nextflow
+    final String dockerArch
+    final String spackArch
+
+    // defined, but currently not used
+    final String platform
+    final String arch
+    final String target
+
+    @CompileStatic
+    protected String getPlatform( String value ) {
+        // return value.minus(~'/.*') // keeping for reference
+        def chunks = value.tokenize('/')
+        if( chunks.size() > 1 )
+            return chunks[0]
+        else
+            return null
+    }
+
+    @CompileStatic
+    protected String getArch( String value ) {
+        // return value.minus(~'.*/') // keeping for reference
+        def chunks = value.tokenize('/')
+        if( chunks.size() == 3 )
+            return chunks[1] + '/' + chunks[2]
+        else if( chunks.size() == 2 )
+            return chunks[1]
+        else
+            return chunks[0]
+    }
+
+    @CompileStatic
+    protected String validateArchToSpackArch( String value ) {
+        if( value == 'x86_64' || value == 'amd64' )
+            return 'x86_64'
+        else if( value == 'aarch64' || value == 'arm64' || value == 'arm64/v8' )
+            return 'aarch64'
+        else if( value == 'arm64/v7' )
+            return null
+        else if( value == 'arm' || value == 'arm/v7' || value == 'arm/7' || value == 'arm/v5' || value == 'arm/5' )
+            return 'arm'
+        else
+            return 'ILLEGAL_ARCH'
```

Better to throw the exception here instead returning an invalid string. I guess you made for testing, but still better to split the test in two and validate ok value and invalid values separately

---

Actually, I did it because I did not want to pass `res.name` as an additional argument to the function.

But I agree with your point, so ... done!